packages/drivers/odsp-driver/src/epochTracker.ts

```diff
@@ -0,0 +1,156 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import assert from "assert";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { createOdspNetworkError, fluidEpochMismatchError, OdspErrorType } from "@fluidframework/odsp-doclib-utils";
+import { fetchAndParseAsJSONHelper, fetchHelper, IOdspResponse } from "./odspUtils";
+import { ICacheEntry, LocalPersistentCacheAdapter } from "./odspCache";
+
+/**
+ * This class is a wrapper around fetch calls. It adds epoch to the request made so that the
+ * server can match it with its epoch value in order to match the version.
+ * It also validates the epoch value received in response of fetch calls. If the epoch does not match,
+ * then it also clears all the cached entries for the given container.
+ */
+export class EpochTracker {
+    private _fluidEpoch: string | undefined;
+    private _hashedDocumentId: string | undefined;
+    constructor(
+        private readonly persistedCache: LocalPersistentCacheAdapter,
+        private readonly logger?: ITelemetryLogger) {
+    }
+
+    public set hashedDocumentId(docId: string | undefined) {
```

Comment:
does it need to be string | undefined (you assert below it can't be undefined).
I'd remove undefined. If in some places we have string | undefined type, I'd rather push casting to string up the stack

Reply:
Actually getter and setter should be of same type. That is why it needs to be string | undefined.